Use SMP/OMP/MCUmgr, remove newtmgr and change OICMGR to use OMP

vrahane opened the pull request:
- Replace newtmgr with SMP and MCUmgr
- Please see https://github.com/apache/mynewt-core/issues/1871 for details
- Make stats, logs and img_mgmt work with MCUmgr
- Make default group commands work with MCUmgr
- Add transient packages for backwards compatibility
- Update apps to have the updated dependencies
- Add syscfgs for MGMT commands Eg: CONFIG_MGMT, STATS_MGMT, LOG_MGMT, etc
- Make OICMGR use OMP from MCUmgr
- Transition from Newtmgr -> SMP
- Remove cborattr and add it to MCUmgr
- Fix builds for log version 2
- As the part of this PR, we are also removing support for `log_fcb_slot1` since it is not really used by the community or any modules in the core repo.
Please refer https://github.com/apache/mynewt-mcumgr/pull/27 for corresponding changes to MCUmgr.
- Moved imgmgr logging to img_mgmt in MCUmgr
- Remove nmgr_uart transport since we do not really need it or use it since smp_shell is there, adding transient pkg for it.

mkiiskila commented on sys/log/full/src/log_cbmem.c:
```diff
@@ -70,27 +73,29 @@ log_cbmem_append_mbuf_body(struct log *log, const struct log_entry_hdr *hdr,
                 .om = om,
             },
         },
         .count = 3,
```
Here too. count is only 2 if LOG_VERSION >= 3

mlaz replied:
Did you mean count is only 2 if LOG_VERSION < 3 ?

mkiiskila replied:
Indeed :)

mlaz replied:
Done.

mkiiskila commented on sys/log/full/src/log_fcb.c:
```diff
@@ -257,9 +258,11 @@ log_fcb_append_body(struct log *log, const struct log_entry_hdr *hdr,
     u8p = body;
 
     memcpy(buf, hdr, LOG_BASE_ENTRY_HDR_SIZE);
+#if MYNEWT_VAL(LOG_VERSION) > 2
```
log_fcb2.c also needs these changes.

mlaz replied:
Changed.

mlaz commented on mgmt/imgmgr/include/imgmgr/imgmgr.h:
```diff
@@ -133,16 +145,31 @@ imgr_get_current_hash(uint8_t *hash, uint16_t hashlen);
+
+/** Callback function pointers */
+typedef struct
+{
```
Fixed.

mlaz commented on mgmt/smp/pkg.yml:
```diff
@@ -0,0 +1,41 @@
+#
+
+pkg.name: mgmt/smp
+pkg.description: Server-side SMP functionality.
```
Changed it to Simple Management Protocol.

mlaz commented on fs/fs/src/fs_priv.h:
```diff
@@ -33,8 +33,8 @@ struct fs_ops *safe_fs_ops_for(const char *fs_name);
-#if MYNEWT_VAL(FS_NMGR)
-int fs_nmgr_init(void);
+#if MYNEWT_VAL(FS_MGMT)
+int fs_mgmt_init(void);
```
Fixed.

mlaz commented on mgmt/imgmgr/include/imgmgr/imgmgr.h:
```diff
@@ -117,6 +80,55 @@ int imgr_my_version(struct image_version *ver);
+ * Returns 2 if slot is empty
+ */
+int
+imgr_read_info(int image_slot, struct image_version *ver, uint8_t *hash,
```
Fixed.

mlaz replied:
fixed

mlaz commented on sys/log/full/src/log_cbmem.c:
```diff
@@ -44,10 +46,11 @@ log_cbmem_append_body(struct log *log, const struct log_entry_hdr *hdr,
         .count = 3,
```
Changed like below.

mkiiskila commented:
Travis build failures contain legitimate issues (at least that's how they seemed to me). I would recommend going through those and addressing them.

kasjer commented on mgmt/oicmgr/src/oicmgr.c:
```diff
@@ -19,244 +19,217 @@
-
-    return 0;
+    /* We need to trim from the back because the head
+     * costains useful information which we do not wast
```
costains spelling error

vrahane replied:
done

kasjer commented on mgmt/oicmgr/include/oicmgr/oicmgr.h:
```diff
@@ -40,12 +46,27 @@ int omgr_oic_process_put(oc_request_t *req, oc_interface_mask_t mask);
+ * @param req   Ptr to the OIC request
+ * @param mask  Interface mask
+ */
+void
```
In headers void should be in the line with function header

vrahane replied:
done

kasjer commented on mgmt/imgmgr/include/imgmgr/imgmgr.h:
```diff
@@ -133,16 +142,29 @@ imgr_get_current_hash(uint8_t *hash, uint16_t hashlen);
+} imgmgr_dfu_callbacks_t;
+
+/**
+ * Register iimage manager callbacks
```
iimage?

vrahane replied:
done

vrahane commented on encoding/tinycbor/include/tinycbor/cbor_mbuf_reader.h:
```diff
@@ -22,6 +22,7 @@
 
 #include "os/mynewt.h"
 #include <tinycbor/cbor.h>
+#include "tinycbor/cbor_mbuf_reader.h"
```
fixed

vrahane commented on test/crash_test/src/crash_mgmt.c:
```diff
@@ -64,17 +74,19 @@ crash_test_nmgr_write(struct mgmt_cbuf *cb)
+
+    os_callout_reset(&mynewt_os_mgmt_crash_callout,
+                     delay_ms * OS_TICKS_PER_SEC / 1000);
+
```
resolved

vrahane commented on sys/log/full/syscfg.yml:
```diff
@@ -97,7 +90,7 @@ syscfg.defs:
             entry in it.
         value: 400
-
+    
```
fixed

vrahane commented on sys/log/full/src/log_fcb.c:
```diff
@@ -714,7 +718,7 @@ log_fcb_copy_entry(struct log *log, struct fcb_entry *entry,
     struct log_entry_hdr ueh;
     char data[LOG_PRINTF_MAX_ENTRY_LEN + LOG_BASE_ENTRY_HDR_SIZE + LOG_IMG_HASHLEN];
-    uint16_t hdr_len;
+    uint16_t hdr_len = 0;
```
updated

vrahane commented on sys/log/full/src/log_fcb.c:
```diff
@@ -225,6 +225,7 @@ log_fcb_append_body(struct log *log, const struct log_entry_hdr *hdr,
 
     fcb_log = (struct fcb_log *)log->l_arg;
     fcb = &fcb_log->fl_fcb;
+    hdr_len = 0;
```
updated

vrahane commented on sys/log/full/include/log/log.h:
```diff
@@ -35,6 +35,18 @@
+#else
+#define LOG_HDR_SIZE 15
+#endif
+
```
removed

vrahane commented on mgmt/smp/transport/smp_shell/src/smp_shell.c:
```diff
@@ -23,18 +23,20 @@
 
-static int
-nmgr_shell_out(struct nmgr_transport *nt, struct os_mbuf *m)
+int
```
done

vrahane commented on mgmt/smp/src/smp.c:
```diff
@@ -0,0 +1,323 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
```
resolved

vrahane commented on mgmt/smp/src/smp.c:
```diff
@@ -0,0 +1,323 @@
+    }
+
+    /* We need to trim from the back because the head
+     * costains useful information which we do not wast
```
resolved

vrahane commented on mgmt/smp/transport/ble/syscfg.yml:
```diff
@@ -6,7 +6,7 @@
 # to you under the Apache License, Version 2.0 (the
 # "License"); you may not use this file except in compliance
 # with the License.  You may obtain a copy of the License at
-#
```
done

vrahane commented on mgmt/smp/src/smp.c:
```diff
@@ -0,0 +1,323 @@
+}
+
+/**
+ * Allocates an mbuf to costain an outgoing response fragmest.
```
done

vrahane commented on mgmt/smp/smp_os/src/smp_os.c:
```diff
@@ -0,0 +1,226 @@
+#include <tinycbor/cbor.h>
+#include <cborattr/cborattr.h>
+
+static int smp_def_console_echo(struct mgmt_ctxt *);
```
done

vrahane commented on mgmt/smp/smp_os/src/smp_os.c:
```diff
@@ -0,0 +1,226 @@
+#include <timepersist/timepersist.h>
+#endif
+
+#include "smp_os/smp_os.h"
```
All the system includes are <> and all the library includes are "". That's a convention we have been following. It was originally that way too.

vrahane commented on mgmt/smp/transport/smp_shell/syscfg.yml:
```diff
@@ -1,11 +1,12 @@
 # "License"); you may not use this file except in compliance
 # with the License.  You may obtain a copy of the License at
-#
+# 
```
done

vrahane commented on apps/sensors_test/pkg.yml:
```diff
@@ -34,7 +34,7 @@ pkg.deps:
     - "@apache-mynewt-core/hw/sensor/creator"
     - "@apache-mynewt-core/sys/reboot"
-    - "@apache-mynewt-core/mgmt/newtmgr/transport/ble"
+    - "@apache-mynewt-core/mgmt/mgmt/transport/ble"
```
changed back

vrahane commented on apps/bleuart/pkg.yml:
```diff
@@ -35,6 +35,6 @@ pkg.deps:
-    - "@apache-mynewt-core/mgmt/newtmgr"
-    - "@apache-mynewt-core/mgmt/newtmgr/transport/ble"
+    - "@apache-mynewt-core/mgmt/mgmt"
+    - "@apache-mynewt-core/mgmt/mgmt/transport/ble"
```
changed

vrahane commented on apps/splitty/pkg.yml:
```diff
@@ -30,8 +30,8 @@ pkg.deps:
-    - "@apache-mynewt-core/mgmt/newtmgr"
-    - "@apache-mynewt-core/mgmt/newtmgr/transport/nmgr_shell"
+    - "@apache-mynewt-core/mgmt/mgmt"
+    - "@apache-mynewt-core/mgmt/mgmt/transport/mgmt_shell"
```
done

vrahane commented on mgmt/smp/transport/smp_uart/src/smp_uart.c:
```diff
@@ -23,47 +23,47 @@
+      (int)&(((struct smp_uart_state *)0)->sus_cb_ev))
+
+struct smp_uart_state {
+    struct smp_transport sus_transport; /* keep first in struct */
```
`smp.h` has the pseudo definition for it.

kasjer replied:
Error: repos/apache-mynewt-core/mgmt/smp/transport/smp_uart/src/smp_uart.c:51:26: error: field 'sus_transport' has incomplete type
     struct smp_transport sus_transport; /* keep first in struct */
                          ^~~~~~~~~~~~~

vrahane replied:
smp_uart removed

vrahane commented on apps/bsnprph/pkg.yml:
```diff
@@ -27,8 +27,9 @@ pkg.deps:
-    - "@apache-mynewt-core/mgmt/newtmgr"
-    - "@apache-mynewt-core/mgmt/newtmgr/transport/ble"
+    - "@apache-mynewt-core/mgmt/mgmt"
+    - "@apache-mynewt-core/mgmt/mgmt/transport/ble"
```
done

vrahane commented on mgmt/smp/transport/smp_uart/src/smp_uart.c:
```diff
@@ -296,56 +296,56 @@ nmgr_uart_rx_pkt(struct nmgr_uart_state *nus, struct os_mbuf_pkthdr *rxm)
+        assert(!sus->sus_rx_q);
+        sus->sus_rx_q = sus->sus_rx;
+        sus->sus_rx = NULL;
+        os_eventq_put(mgmt_evq_get(), &sus->sus_cb_ev);
```
added in smp.h

kasjer replied:
repos/apache-mynewt-core/mgmt/smp/transport/smp_uart/src/smp_uart.c:354:23: error: implicit declaration of function 'mgmt_evq_get'; did you mean 'mgmt_evt'? [-Werror=implicit-function-declaration]
         os_eventq_put(mgmt_evq_get(), &sus->sus_cb_ev);
                       ^~~~~~~~~~~~
                       mgmt_evt

vrahane replied:
removed smp_uart

vrahane commented on mgmt/mgmt/include/mgmt/mgmt.h:
```diff
@@ -1,128 +0,0 @@
- * fewer. This 127-byte maximum applies to the entire frame, including header,
- * CRC, and terminating newline.
- */
-#define MGMT_NLIP_MAX_FRAME     127
```
added in the smp_uart.c source file

vrahane commented on sys/metrics/pkg.yml:
```diff
@@ -27,7 +27,7 @@ pkg.keywords:
     - kernel/os
     - sys/log/full
-    - encoding/tinycbor
+    - '@apache-mynewt-core/encoding/tinycbor'
```
fixed

vrahane commented on mgmt/smp/src/smp.c:
```diff
@@ -0,0 +1,323 @@
+        goto err;
+    }
+
+    return (0);
```
done

vrahane commented:
@mkiiskila thanks, I changed CI a bit to work with mcumgr and fixed the unit tests and other build issues. The build now succeeds, after the PR gets merged in I will change the CI to use master instead.

@kasjer thanks, Miguel and I fixed all the issues you mentioned above, thanks for catching them.

kasjer commented on mgmt/smp/transport/smp_uart/src/smp_uart.c:
```diff
@@ -354,42 +360,42 @@ nmgr_uart_rx_char(void *arg, uint8_t data)
+    struct smp_uart_state *sus = &smp_uart_state;
     int rc;
     struct uart_conf uc = {
         .uc_speed = MYNEWT_VAL(NMGR_UART_SPEED),
```
bin/targets/sensors_test/generated/include/syscfg/syscfg.h:15:49: error: 'MYNEWT_VAL_NMGR_UART_SPEED' undeclared (first use in this function); did you mean 'MYNEWT_VAL_SMP_UART_SPEED'?
 #define MYNEWT_VAL(_name)                       MYNEWT_VAL_ ## _name
                                                 ^
repos/apache-mynewt-core/mgmt/smp/transport/smp_uart/src/smp_uart.c:376:21: note: in expansion of macro 'MYNEWT_VAL'
         .uc_speed = MYNEWT_VAL(NMGR_UART_SPEED),
                     ^~~~~~~~~~

vrahane replied:
removed smp_uart

mkiiskila replied:
This was for low memory platforms which could not fit shell/did not want shell. I've used it, I would have liked to keep it.

kasjer commented on mgmt/smp/transport/smp_uart/src/smp_uart.c:
```diff
@@ -354,42 +360,42 @@ nmgr_uart_rx_char(void *arg, uint8_t data)
 
-    nus->nus_dev =
+    sus->sus_dev =
       (struct uart_dev *)os_dev_open(MYNEWT_VAL(NMGR_UART), 0, &uc);
```
NMGR_UART name is change I guess to SMP_UART

vrahane replied:
removed smp_uart

kasjer commented on mgmt/smp/transport/smp_uart/src/smp_uart.c:
```diff
@@ -354,42 +360,42 @@ nmgr_uart_rx_char(void *arg, uint8_t data)
     SYSINIT_ASSERT_ACTIVE();
 
-    rc = nmgr_transport_init(&nus->nus_transport, nmgr_uart_out, nmgr_uart_mtu);
+    rc = smp_transport_init(&sus->sus_transport, smp_uart_out, smp_uart_mtu);
```
repos/apache-mynewt-core/mgmt/smp/transport/smp_uart/src/smp_uart.c:389:50: error: passing argument 2 of 'smp_transport_init' from incompatible pointer type [-Werror=incompatible-pointer-types]
     rc = smp_transport_init(&sus->sus_transport, smp_uart_out, smp_uart_mtu);
                                                  ^~~~~~~~~~~~
In file included from repos/apache-mynewt-core/mgmt/smp/transport/smp_uart/src/smp_uart.c:26:0:
repos/apache-mynewt-core/mgmt/smp/include/mynewt_smp/smp.h:66:5: note: expected 'smp_transport_out_func_t {aka int (*)(struct os_mbuf *)}' but argument is of type 'int (*)(struct smp_transport *, struct os_mbuf *)'
 int smp_transport_init(struct smp_transport *st,
     ^~~~~~~~~~~~~~~~~~

vrahane replied:
removed smp_uart

kasjer commented on sys/log/full/src/log_fcb.c:
```diff
@@ -200,7 +200,7 @@ log_fcb_hdr_body_bytes(uint8_t align, uint8_t hdr_len)
     /* Assume power-of-two alignment for faster modulo calculation. */
     assert((align & (align - 1)) == 0);
-
+    
```
this patch added trailing space in 8 places

kasjer commented on mgmt/smp/transport/smp_uart/src/smp_uart.c:
```diff
@@ -23,47 +23,53 @@
 
 #include <mgmt/mgmt.h>
-#include <newtmgr/newtmgr.h>
+#include <smp/smp.h>
```
if `smp/smp.h` was `mynewt_smp/smp.h` some of the build errors would be gone

vrahane replied:
removed smp_uart

vrahane commented:
Thanks for the comments @kasjer, not sure why I didn’t see these while building or in the CI. Anyways, I’ll make sure that the builds are fixed.

vrahane commented:
@kasjer what do you use it for ?

kasjer commented:
> @kasjer what do you use it for ?

@vrahane If I understand you question, correctly I have sensors_test application with uart as transport.

vrahane commented:
@kasjer I did speak to @ccollins476ad about it, we do have the shell transport which works over for smp over serial. We do not need smp_uart anymore, I am going to remove it.

vrahane commented:
@kasjer, so, I think all your concerns are now resolved, right ?

vrahane commented:
@mkiiskila awaiting your approval as well, no more errors on CI.

vrahane commented:
@mkiiskila Re-adding smp_uart and @kasjer fixing build.
